python/ray/_private/runtime_env/uri_cache.py

```diff
@@ -0,0 +1,105 @@
+import logging
+from typing import Dict, Callable
+
+default_logger = logging.getLogger(__name__)
+
+DEFAULT_MAX_URI_CACHE_SIZE_BYTES = (1024**3) * 10  # 10 GB
+
+
+class URICache:
+    """
+    Caches URIs up to a specified total size limit.
+
+    URIs are represented by strings.  Each URI has an associated size on disk.
+
+    When a URI is added to the URICache, it is marked as "in use".
+    When a URI is no longer in use, the user of this class should call
+    `mark_unused` to signal that the URI is safe for deletion.
+
+    URIs in the cache can be marked as "in use" by calling `mark_used`.
+
+    Deletion of URIs on disk does not occur until the size limit is exceeded.
+    When this happens, URIs that are not in use are deleted randomly until the
+    size limit is satisfied, or there are no more URIs that are not in use.
+
+    It is possible for the total size on disk to exceed the size limit if all
+    the URIs are in use.
+
+    """
+
+    def __init__(
+            self,
+            delete_fn: Callable[[str, logging.Logger], int] = lambda x, y: 0,
+            max_total_size_bytes: int = DEFAULT_MAX_URI_CACHE_SIZE_BYTES,
+    ):
+        # Maps URIs to the size in bytes of their corresponding disk contents.
+        self._used_uris: Dict[str, int] = dict()
+        self._unused_uris: Dict[str, int] = dict()
+        self._delete_fn = delete_fn
+        self.max_total_size_bytes = max_total_size_bytes
+
+    def mark_unused(self, uri: str, logger: logging.Logger = default_logger):
+        """Mark a URI as unused and okay to be deleted."""
+        if uri not in self._used_uris:
+            logger.debug(f"URI {uri} is already unused.")
+        else:
+            self._unused_uris[uri] = self._used_uris[uri]
+            del self._used_uris[uri]
+        logger.debug(f"Marked uri {uri} unused.")
+        self._evict_if_needed(logger)
+        self._check_valid()
+
+    def mark_used(self, uri: str, logger: logging.Logger = default_logger):
+        """Mark a URI as in use.  URIs in use will not be deleted."""
+        if uri in self._used_uris:
+            return
+        elif uri in self._unused_uris:
+            self._used_uris[uri] = self._unused_uris[uri]
+            del self._unused_uris[uri]
+        else:
+            raise ValueError(f"Got request to mark URI {uri} unused, but this "
+                             "URI is not present in the cache.")
+        logger.debug(f"Marked URI {uri} used.")
+        self._check_valid()
+
+    def add(self,
+            uri: str,
+            size_bytes: int,
+            logger: logging.Logger = default_logger):
+        """Add a URI to the cache and mark it as in use."""
+        if uri in self._unused_uris:
+            if size_bytes != self._unused_uris[uri]:
+                logger.warning(f"Added URI {uri} with size {size_bytes}, which"
```

Comment:
Is this warning really useful? Is it a bug that size bytes are changed?

Reply:
URIs are content hashes, so we don't expect the size to change.  Let me change this to a debug log and add a comment explaining it

Reply:
Can we just assert in this case?

Reply:
Also, I feel like if the content size is not expected to be changed. Do we even need to track of size of used/unused uris?

Why don't we make both used/unused as a set and keep track of the total uri sizes in a separate integer attribute?

Reply:
Feel like this will make code a lot simpler

Reply:
If we only keep track of one integer `total_uri_size`, then how can we update `total_uri_size` when changing a particular URI from used to unused, or vice versa?  Don't we need to store a map from URIs to their sizes?

Reply:
> Why don't we just assert in this case?

That sounds fine to me.

Reply:
> If we only keep track of one integer total_uri_size, then how can we update total_uri_size when changing a particular URI from used to unused, or vice versa? Don't we need to store a map from URIs to their sizes?

It seems like the current API only cares the sum of used + unused. And it is guaranteed one uri is only in one of them; doesn't this mean we don't need this -> `update total_uri_size when changing a particular URI from used to unused`? I might be missing something though

Reply:
yes that's right! So we can remove this per-uri size tracking logic since it doesn't seem to be unnecessary (even after we introduce LRU)